master/internal/api_token.go

```diff
@@ -0,0 +1,215 @@
+package internal
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/uptrace/bun"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/determined-ai/determined/master/internal/api"
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/grpcutil"
+	"github.com/determined-ai/determined/master/internal/token"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/proto/pkg/apiv1"
+)
+
+// PostAccessToken takes user id and optional lifespan, description and creates an
+// access token for the given user.
+func (a *apiServer) PostAccessToken(
+	ctx context.Context, req *apiv1.PostAccessTokenRequest,
+) (*apiv1.PostAccessTokenResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	targetFullUser, err := getFullModelUser(ctx, model.UserID(req.UserId))
+	if err != nil {
+		return nil, err
+	}
+	targetUser := targetFullUser.ToUser()
+	if err = token.AuthZProvider.Get().CanCreateAccessToken(ctx, *curUser, targetUser); err != nil {
+		return nil, status.Error(codes.PermissionDenied, err.Error())
+	}
+
+	tokenExpiration := token.DefaultTokenLifespan
+	if req.Lifespan != "" {
+		d, err := time.ParseDuration(req.Lifespan)
+		if err != nil || d < 0 {
+			return nil, status.Error(codes.InvalidArgument,
+				"Lifespan must be a Go-formatted duration string with a positive value")
+		}
+		tokenExpiration = d
+	}
+
+	token, tokenID, err := token.CreateAccessToken(
+		ctx, targetFullUser.ID, token.WithTokenExpiry(&tokenExpiration), token.WithTokenDescription(req.Description))
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.PostAccessTokenResponse{Token: token, TokenId: int32(tokenID)}, nil
+}
+
+type accessTokenFilter struct {
```

Comment:
why not just make this a part of the request? as in, make have these arguments as part of the request object. this is confusing, and introduces unnecessary parsing both client and server side.

Reply:
Earlier we did have these as a separate input to the API request. But after the discussion (will have to look for it, couldn't find this minute), we agreed to follow the norm of how `/api/v1/runs` endpoint does it. 

@corban-beaird : thoughts?

Reply:
The goal had been to be aligned with the SearchRuns/SearchExperiments APIs, so that when we do move towards creating a webUI token management page that it would be a smoother transition to implement the latest table framework. 

However, if that's not a priority, I agree that this approach hides what the valid filters are and requires a bit of digging to resolve; it's worth roping in @johnkim-det to get his input on whether or not he's thinks it would be more difficult to create a token management page if we don't follow the same structure as SearchRuns & Experiments and instead have the request body have explicit fields for each of these.

The down side is that if we every have any sort of dynamic fields we want to filter on for tokens this will be more difficult to pull of with explicitly defined arguments; however, since we're not looking at that for the near future, I'm not opposed to making the transition away from the filter string if it's not an undue hurdle for the webUI

Reply:
Don't think there's a reason to have this use a filter string to match SearchRuns/SearchExperiments.

We need that specifically for the way filtering works on Runs/Experiments List, but all other tables (users, workspaces, models, etc.) have filtering using top-level fields. Even if we change these to use the same table component as Runs/Experiments List, that would not require using the same filtering strategy.